eng/versioning.targets

```diff
@@ -157,9 +157,10 @@
   </Target>
 
   <!-- Removes assembly level attributes for Browser. -->
-  <Target Name="RemoveSupportedPlatformAssemblyAttributeForBrowser" AfterTargets="GetAssemblyAttributes" Condition="'$(TargetFrameworkSuffix)' == 'Browser'">
+  <Target Name="RemoveSupportedPlatformAssemblyAttributeForBrowser" AfterTargets="GetAssemblyAttributes" Condition="'$(TargetFrameworkSuffix)' == 'Browser' or '$(IsTestProject)' == 'true'">
```

Comment:
There is build error on test projects, i believe we don't want/need to add attributes to the test build, but we will need to keep the rule for that. Anyway i was not able to figure out how to fix this error, please suggest
`artifacts\obj\Microsoft.Extensions.Logging.Tests\net6.0-windows-Debug\Microsoft.Extensions.Logging.Tests.AssemblyInfo.cs(23,38): error CS0433: (NETCORE_ENGINEERING_TELEMETRY=Build) The type 'TargetPlatformAttribute' exists in both 'Microsoft.Extensions.Logging.EventLog, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' and 'System.Runtime, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'`

`artifacts\obj\Microsoft.Extensions.Logging.Tests\net6.0-windows-Debug\Microsoft.Extensions.Logging.Tests.AssemblyInfo.cs(24,38): error CS0433: (NETCORE_ENGINEERING_TELEMETRY=Build) The type 'SupportedOSPlatformAttribute' exists in both 'Microsoft.Extensions.Logging.EventLog, Version=6.0.0.0, Culture=neutral, PublicKeyToken=adb9793829ddae60' and 'System.Runtime, Version=6.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'`

Reply:
We should stay on the common path where possible and not disable default behavior that comes via the sdk without a reason... For this specific build issue I'll defer to @ericstj. Maybe he knows just by looking at the failure. I would need to look at a binlog to know more.

Reply:
@buyaa-n and I dug into this and here's what we found:

* Microsoft.Extensions.Logging.EventLog has `<IncludePlatformAttributes>true</IncludePlatformAttributes>` so that it can apply `[SupportedOSPlatform("windows")]` when targeting `netstandard2.0`
* That project also has `InternalsVisibleTo` the test project
* The test project builds against `NetCoreAppCurrent`, referencing the EventLog assembly
* The test project therefore has visibility to 2 copies of the attribute types--1 in corelib, and 1 internal in the EventLog assembly

I agree with @ViktorHofer that we should have as few exceptions to the SDK tooling as possible. This problem will be generalized such that:

1. A downlevel target that internally defines the platform attributes
2. That also has internals visible to another project
3. The other project would be unable to use the platform attributes

I don't know if we have any similar challenges elsewhere or if there's an existing approach we can follow to overcome this.

Reply:
Here's what we're seeing:

1. The SDK [will always add](https://github.com/dotnet/sdk/blob/92b9068a325295a385006a4cad36986a275c4622/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateAssemblyInfo.targets#L128-L138) the `[SupportedOSPlatform]` attribute when targeting a specific platform; we don't see a way to disable the generation of the attribute.
2. Since this error is produced inside the assemblyinfo.cs file, there likely wouldn't be a way to use an `extern alias` to disambiguate the type reference to the attribute
3. We can't disable `InternalsVisibleTo` either

With this, we don't see a clear path to keep the attribute in place in test projects that cover `netstandard2.0` builds that use `<IncludePlatformAttributes>` and `InternalsVisibleTo` or to prevent the attribute from getting generated. That leaves us with having to _remove_ the attribute like is being done here.

@ViktorHofer / @ericstj -- would you prefer we invoke this target to remove the attribute only from the `Extensions.Logging` test project instead of doing it here for all test projects? Let us know if you have any other suggestions.

@buyaa-n Please file an issue against the SDK that we should have an escape hatch to prevent the `[SupportedOSPlatform]` attribute from being added for these types of edge cases.

Reply:
Great findings @buyaa-n and @jeffhandley.

> The SDK will always add the [SupportedOSPlatform] attribute when targeting a specific platform; we don't see a way to disable the generation of the attribute.

Which part of our system adds the `SupportedOSPlatform` attribute types? Is it the SDK or is that dotnet/runtime specific? If the latter, should we just add a escape hatch condition?

> We can't disable InternalsVisibleTo either

I agree. Even though InternalsVisibleTo isn't without concerns, our infra should not forbid using it.

Reply:
> Which part of our system adds the SupportedOSPlatform attribute types? Is it the SDK or is that dotnet/runtime specific? If the latter, should we just add a escape hatch condition?

The attribute added in AssemblyInfo.cs is added by [SDK](https://github.com/dotnet/sdk/blob/92b9068a325295a385006a4cad36986a275c4622/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateAssemblyInfo.targets#L128-L138) (We also add dotnet/runtime specific attributes in ref/src, but that is not the issue here )

Reply:
It's in the SDK when targeting a specific platform through the TFM.

https://github.com/dotnet/sdk/blob/92b9068a325295a385006a4cad36986a275c4622/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.GenerateAssemblyInfo.targets#L128-L138

Reply:
Sorry, let me rephrase. Who adds the internal types?

Reply:
We add it to make the attributes visible for targets < 5.0 https://github.com/dotnet/runtime/blob/b45ee9d37afec0c88141053e86ccf71c6f283000/src/libraries/System.Private.CoreLib/src/System/Runtime/Versioning/PlatformAttributes.cs#L31-L36  that test project references projects having references to both

Reply:
Just chiming in here.  This is really only a problem in usage from a project that has IVT to libs that define this internally.  I don't have a huge preference for how we solve it.  We could let that lib define a local copy (local types are preferred over referenced without error) or just remove the attribute.  Does the attribute application even make any sense on a test assembly?  Who consumes this?

Reply:
i think nothing consuming it, as far as I know netcore analyzers are currently disabled in test projects including CA1416 https://github.com/dotnet/runtime/blob/c2404a90b585babebd466d6c6257b763f06f76a0/eng/CodeAnalysis.test.ruleset#L67

Reply:
I don't see much value in having the SDK add this attribute for test projects; this could be a common pitfall since InternalsVisibleTo is often needed for test projects.

Using the target to remove the attribute from all test projects in the runtime feels OK to me (while waiting for the SDK to stop generating the attribute for test projects).

Reply:
There is no any update on [the issue filed to SDK for an option stopping generating the attribute](https://github.com/dotnet/sdk/issues/14836), maybe they are out for the rest of the year, anyway i think that should not a block this PR, this is needed for the 2nd step of the effort `Fix browser build warnings and enable TFM Supported attribute for browser`. Could we merge this PR? Is anyone has any other question or there anything i am missing? 
cc @jeffhandley @ericstj @ViktorHofer @joperezr @Anipik

Reply:
> There is no any update on the issue filed to SDK for an option stopping generating the attribute, maybe they are out for the rest of the year, anyway i think that should not a block this PR

I agree. Especially as we don't want to upgrade to a newer SDK just for this feature.

Reply:
I'm fine with the PR as-is.